dora/src/cannon/codegen.rs

```diff
@@ -938,6 +990,191 @@ where
             self.asm.bind_label_to(jump.label, offset);
         }
     }
+
+    fn emit_new_object(&mut self, dest: Register, class_def_id: ClassDefId) {
+        assert_eq!(self.bytecode.register_type(dest), BytecodeType::Ptr);
+
+        let cls = self.vm.class_defs.idx(class_def_id);
+        let cls = cls.read();
+
+        self.asm.emit_comment(Comment::Alloc(class_def_id));
+
+        let alloc_size = match cls.size {
+            InstanceSize::Fixed(size) => AllocationSize::Fixed(size as usize),
+            _ => unimplemented!(
+                "class size type {:?} for new object not supported",
+                cls.size
+            ),
+        };
+
+        let array_ref = match cls.size {
+            InstanceSize::ObjArray => true,
+            _ => false,
+        };
+
+        let gcpoint = GcPoint::from_offsets(self.references.clone());
+        self.asm.allocate(
+            REG_RESULT.into(),
+            alloc_size,
+            Position { line: 0, column: 0 }, // Correct position is currently not known.
+            array_ref,
+            gcpoint,
+        );
+
+        let bytecode_type = self.bytecode.register_type(dest);
+        let offset = self.bytecode.register_offset(dest);
+
+        // store gc object in temporary storage
+        self.asm
+            .store_mem(bytecode_type.mode(), Mem::Local(offset), REG_RESULT.into());
+
+        // store classptr in object
+        let cptr = (&**cls.vtable.as_ref().unwrap()) as *const VTable as *const u8;
+        let disp = self.asm.add_addr(cptr);
+        let pos = self.asm.pos() as i32;
+
+        self.asm.emit_comment(Comment::StoreVTable(class_def_id));
+        self.asm.load_constpool(REG_TMP1.into(), disp + pos);
+        self.asm
+            .store_mem(MachineMode::Ptr, Mem::Base(REG_RESULT, 0), REG_TMP1.into());
+
+        // clear mark/fwdptr word in header
+        assert!(Header::size() == 2 * mem::ptr_width());
+        self.asm.load_int_const(MachineMode::Ptr, REG_TMP1, 0);
+        self.asm.store_mem(
+            MachineMode::Ptr,
+            Mem::Base(REG_RESULT, mem::ptr_width()),
+            REG_TMP1.into(),
+        );
+
+        match cls.size {
```

Comment:
Should always be `Fixed`

Reply:
Removed the match and instead use `alloc_size` from above.

Reply:
Added a unreachable, but I guess there is a better way. I will look into this.

Reply:
Usually I add methods to the struct for that (here it would be `to_fixed()`)

Reply:
But let's keep it this way in this PR